fix: copy dependencies on aliasing to avoid sharing chart references on multiply aliased dependencies

dastrobu opened the pull request:
Dependencies keep a reference on their parent chart, which breaks if a chart reference is shared among multiple aliases.
By copying the dependencies, parent information can be set correctly to render the templates as expected later on.

Note that this change will make ChartFullPath return a different path for sub-subcharts. It will contain the alias names instead of the path to the chart files which makes it consistent with paths to templates on the subchart level.

Closes #9150

Signed-off-by: Daniel Strobusch <1847260+dastrobu@users.noreply.github.com>

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

dastrobu committed:
copy dependencies on aliasing to avoid sharing chart references on multiply aliased dependencies

Dependencies keep a reference on their parent chart, which breaks if a chart reference is shared among multiple aliases.
By copying the dependencies, parent information can be set correctly to render the templates as expected later on.

Note that this change will make ChartFullPath return a different path for sub-subcharts. It will contain the alias names instead of the path to the chart files which makes it consistent with paths to templates on the subchart level.

Closes #9150

Signed-off-by: Daniel Strobusch <1847260+dastrobu@users.noreply.github.com>

itscaro commented:
@dastrobu Thank you for the PR, this is exactly what I need 👍

dastrobu commented:
@itscaro thanks for the feedback. Just need to get the maintainers attention to get someone reviewing this PR.

whirm commented:
Bump! I'm also very interested on getting it merged.

thefang12 commented:
Bump! this would greatly simplify my charts 🙏

Ornias1993 commented:
@bacongobbler you might have someone at least... well... look at this?

tavlima commented:
Beg your pardon, @bacongobbler, but is there any chance this can be released as a patch version in the 3.7.x train, instead of waiting for 3.8.0? This bug makes it really hard for one to create reusable/composable charts...

trigger305 commented:
would also like this one please as part of 3.8.2

mederel commented on pkg/chartutil/dependencies.go:
```diff
@@ -106,6 +107,14 @@ func getAliasDependency(charts []*chart.Chart, dep *chart.Dependency) *chart.Cha
 
+		// empty dependencies and shallow copy all dependencies, otherwise parent info may be corrupted if
+		// there is more than one dependency aliasing this chart
+		out.SetDependencies()
```
I seems from my local testing and the usecase I have (attached in zip) this is not enough while the following changes from this [other PR](https://github.com/helm/helm/pull/10174/files) is working:

```golang
		if md.Dependencies != nil {
			dependencies := make([]*chart.Dependency, len(md.Dependencies))
			for i := range md.Dependencies {
				d := *md.Dependencies[i]
				dependencies[i] = &d
			}
			md.Dependencies = dependencies
		}
```

[reproducer-common-dep-import-value.zip](https://github.com/helm/helm/files/8602843/reproducer-common-dep-import-value.zip)

dastrobu replied:
I am happy to look into this, when there is feedback from the maintainers what milestone is realistic for this PR. If there is a case I overlooked, I guess there should be another test case as well.

dastrobu replied:
@mederel since I had some spare time and these issues are relly quite annoying, I took a look at the issue you
mentioned.

The example from your zip boils down to the following graph:

```mermaid
graph TD;
    umbrella-->child1;
    umbrella-->child2;
    umbrella-->child2-bis;
    child2-bis-. alias .->child2;
    child1-->common;
    child2-->common;
    child2-bis-->common;
```

The example can be a bit simplified by removing `child1`, while still showing the issue

```mermaid
graph TD;
    umbrella-->child2;
    umbrella-->child2-bis;
    child2-bis-. alias .->child2;
    child2-->common;
    child2-bis-->common;
```

So we are essentially back to the test case from #9150, which is

```mermaid
graph TD;
    parent-->foo;
    parent-->bar;
    foo-. alias .->child;
    bar-. alias .->child;
    foo-->grandchild;
    bar-->grandchild;
```

The only notably difference is the use of `import-values` in the two child charts and this is where it gets interesting.

The issue reported in #9150 was related to information stored in
https://github.com/helm/helm/blob/ced54b12896ae933017782e2bbd722e2bc8c0914/pkg/chart/chart.go#L55
which is why I copied the `Chart` objects.

The issue you found is similar but in a different place:
https://github.com/helm/helm/blob/ced54b12896ae933017782e2bbd722e2bc8c0914/pkg/chart/dependency.go#L47
which is why you copied the `Dependency` objects in the metadata.

So im the end we need both. I updated this PR with a fix and a test case for the issue you mentioned.
I have to say that I don't like the solution. There seems to be a conceptual issue with aliases in Helm's dependency
management implementation from my point of view.

Some parts of the implementation seem to assume to reflect alias with the following graph in memory:

```mermaid
graph TD;
    parent-->child;
    parent-- aliased -->child;
```

while others assume

```mermaid
graph TD;
    parent-->child;
    parent-- aliased --> child-copy;
```

Depending on which path to follow information such as "imported values" or "parent" must be stored on the edges of the graph (first diagram) or can be stored on the nodes (second diagram). 

Anyway, there is no easy and nice way to fix this, but I would like to see a refactoring of the logic for Helm 4.

prune998 commented:
are Helm maintainers dead ? How many follower does it take to get on the top of the stack to be reviewed ? 
As far as I have tested, I think the [example in the doc](https://helm.sh/docs/topics/charts/#alias-field-in-dependencies) would not work

mprocterncs commented:
I'm currently waiting for this fix.  Will it be delivered into 3.10.0?

stavros-k commented:
Maybe 3.10.2? :)

Ornias1993 commented:
@mattfarina and @hickeyma  Instead of just postponing can someone at least try to pretent putting a second of effort into community PR's like these?

joejulian commented:
disclaimer: I'm not a core maintainer

Has anyone else tried this patch? If you can build, test, and report back your results with your grey :heavy_check_mark:, I've heard that would help.

pepesenaris commented:
> disclaimer: I'm not a core maintainer
> 
> Has anyone else tried this patch? If you can build, test, and report back your results with your grey heavy_check_mark, I've heard that would help.

I tried them locally.
Built locally after merging master into the branch, see attached zip. Worked as expected, when with the 3.10+bugfix binary we got 2 different sets of resources, one for each grandchild that was using the same dependency with different values.
[helm-3.zip](https://github.com/helm/helm/files/9990942/helm-3.zip)

pepesenaris commented:
@mattfarina What would the team need to move this forward? How can I help?

iscultas commented:
2 years...

dastrobu committed:
copy dependency metadata on aliasing to avoid sharing imported values

imported values are stored in dependency objects, which breaks if a chart dependency is shared among multiple aliases.
By copying the dependency objects in the metadata values can be imported correctly.

Supersedes #10174

Signed-off-by: Daniel Strobusch <1847260+dastrobu@users.noreply.github.com>

dastrobu commented:
The PR was rebased and updated with another test case as well as a fix mentioned in https://github.com/helm/helm/pull/9175#discussion_r862808962. I think not only I, but a significant part of the Helm community would love to see feedback from the maintainers.

lucasrod16 reviewed (approved):
Note: I am not a maintainer.

I was able to reproduce this bug with the steps provided in the [issue](https://github.com/helm/helm/issues/9150). The changes introduced in this PR resolve the issue and allows helm to template out the aliased dependencies correctly. Test cases have been added to test this behavior and workflows are passing. LGTM

lucasrod16 commented:
I'd like to share a few options to help potentially bump PR reviews in the future:
- `helm-dev` channel in [Kubernetes slack](https://kubernetes.slack.com/messages/helm-dev)
- [weekly helm developer meeting on zoom](https://zoom.us/j/696660622?pwd=MGsraXZ1UkVlTkJLc1B5U05KN053QT09)
- [more info](https://github.com/lucasrod16/helm/tree/main#community-discussion-contribution-and-support)

I realize this may not be feasible for some for various reasons, but they are ways that the community can discuss things with each other, and with the maintainers.

joejulian reviewed (dismissed):
This looks good. I think I've seen a bug before that this should fix. I wonder if I can find it again. :grin:

taliastocks commented:
Your commitment is impressive, would love to see this merged :)

dastrobu commented:
> [joejulian](https://github.com/joejulian) added the [Has One Approval](https://github.com/helm/helm/labels/Has%20One%20Approval)  label [on Aug 1](https://github.com/helm/helm/pull/9175#event-9970691081)

What would be the next step?

scottrigby commented:
> > [joejulian](https://github.com/joejulian) added the [Has One Approval](https://github.com/helm/helm/labels/Has%20One%20Approval)  label [on Aug 1](https://github.com/helm/helm/pull/9175#event-9970691081)
> 
> What would be the next step?

Hi @dastrobu are you able to resolve the conflicts in this PR so we can get this merged? I also pinged you in k8s slack `#helm-dev` channel [here](https://kubernetes.slack.com/archives/C51E88VDG/p1744835831146189). Thanks!

dastrobu commented:
@scottrigby thanks for coming back to this. I rebased and updated the PR and resolved the issues. It should be ready for review now.

scottrigby reviewed (approved):
@dastrobu thanks for your contribution and persistence.
manually tested, has unit tests, and code review LGTM 👍

scottrigby commented:
Removed `3.18.0` milestone, because the `main` branch is now for Helm 4.

@dastrobu are you able to make a v3 backport PR with `dev-v3` as the base branch? That PR will be assigned the appropriate milestone. See [Maintaining Helm 3](https://github.com/helm/community/blob/main/hips/hip-0012.md#maintaining-helm-3) section of the Helm 4 Development Process HIP for more info on this.

dastrobu commented:
@scottrigby thanks for the explanation, I did miss that helm 4 is already under development. Find the back port for helm 3 in #30802.

pepesenaris commented:
Thanks everyone... We've been tracking this PR for 2 years. Great see it merged
